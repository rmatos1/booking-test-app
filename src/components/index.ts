export * from "./bookingDescription";
export * from "./bookingForm";
export * from "./images";
export * from "./inputGroup";
export * from "./navBar";
export * from "./styledComponents";
export * from "./testComponent";
