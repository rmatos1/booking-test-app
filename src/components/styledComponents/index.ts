export * from './buttons';
export * from './formComponents';
export * from './images';
export * from './layouts';
export * from './lists';
export * from './modalComponents';
export * from './texts';
