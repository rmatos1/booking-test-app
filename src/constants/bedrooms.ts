import { images } from '../assets';

export interface IBedroom {
  number: number;
  type: string;
  beds: {
    double?: number;
    single?: number;
  };
  img: string;
  fullCapacity: number;
  dailyPrice: number;
}

/**
 * constant that stores the maximum number of people possible in the biggest bedroom
 */
export const MAX_CAPACITY_BEDROOM = 4;

/**
 * constant with the available bedrooms to be booked
 */
export const hotelBedrooms: IBedroom[] = [
  {
    number: 1,
    type: 'Quadruple',
    beds: {
      double: 1,
      single: 2,
    },
    img: images.bedroom1,
    fullCapacity: 4,
    dailyPrice: 65,
  },
  {
    number: 2,
    type: 'Single',
    beds: {
      double: 1,
    },
    img: images.bedroom2,
    fullCapacity: 2,
    dailyPrice: 40,
  },
  {
    number: 3,
    type: 'Triple',
    beds: {
      double: 1,
      single: 1,
    },
    img: images.bedroom3,
    fullCapacity: 3,
    dailyPrice: 55,
  },
  {
    number: 4,
    type: 'Triple',
    beds: {
      single: 3,
    },
    img: images.bedroom4,
    fullCapacity: 3,
    dailyPrice: 50,
  },
  {
    number: 5,
    type: 'Single',
    beds: {
      single: 2,
    },
    img: images.bedroom5,
    fullCapacity: 2,
    dailyPrice: 35,
  },
];
