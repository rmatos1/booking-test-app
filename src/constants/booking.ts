export const initialBookingData = {
  checkIn: '',
  checkOut: '',
  qtyGuests: '',
  name: '',
  email: '',
  selectedBedroom: 0,
};

export const testBookingData = {
  id: '1',
  name: 'Julia',
  email: 'julia@teste.com',
  checkIn: '2023-12-19',
  checkOut: '2023-12-26',
  selectedBedroom: 3,
  qtyGuests: 3,
  totalPrice: 450,
};
