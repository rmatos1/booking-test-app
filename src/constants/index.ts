export * from './bedrooms';
export * from './booking';
export * from './styles';
