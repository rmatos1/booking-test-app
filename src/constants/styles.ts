/*
 * break points
 */
export const SMALL_DESKTOP = '990px';
export const TABLET = '768px';
export const MOBILE = '460px';

/***************************** */

export const HORIZONTAL_PADDING_DESKTOP = '25px';

export const HORIZONTAL_PADDING_MOBILE = '15px';
