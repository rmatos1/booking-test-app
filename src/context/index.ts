export * from './bookingContext.provider';
export * from './confirmedBookingsContext.provider';
export * from './drawerContext.provider';
