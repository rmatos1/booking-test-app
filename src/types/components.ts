import { ReactNode } from "react";

export interface IComponentWithChildren {
  children?: ReactNode;
}

export interface IBaseDrawer extends IComponentWithChildren {
  isDrawerVisible: boolean;
  onCloseDrawer: () => void;
}
