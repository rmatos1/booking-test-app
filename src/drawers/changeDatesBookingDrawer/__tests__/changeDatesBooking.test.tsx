import { describe, test } from '@jest/globals';
import { render, fireEvent } from '@testing-library/react';
import { TestComponent } from '../../../components';
import { testBookingData } from '../../../constants';
import { IConfirmedBookingsContext } from '../../../context';
import moment from "moment"

import { ChangeDatesBookingDrawer } from '..';

const defaultConfirmedBookingsContextValue = {
  confirmedBookings: [],
  setConfirmedBookings: jest.fn(),
}

const defaultProps = {
  bookingData: testBookingData,
};

const setup = (confirmedBookingsContextValue?: IConfirmedBookingsContext): JSX.Element => {
  
  return (
    <TestComponent confirmedBookingsContextValue={confirmedBookingsContextValue || defaultConfirmedBookingsContextValue}>
      <ChangeDatesBookingDrawer {...defaultProps} />
    </TestComponent>
  );
};

describe('<ChangeDatesBookingDrawer />', () => {
  
  test('should render correctly', () => {
    const wrapper = render(setup());

    expect(wrapper).toMatchSnapshot();
  });

  test('should render the defined check-in value', () => {
    const wrapper = render(setup());

    const inputCheckIn = wrapper.getByTestId('input-check-in') as HTMLInputElement;

    expect(inputCheckIn.value).toBe(defaultProps.bookingData.checkIn);
  });

  test('should render the defined check-out value', () => {
    const wrapper = render(setup());

    const inputCheckOut = wrapper.getByTestId('input-check-out') as HTMLInputElement;

    expect(inputCheckOut.value).toBe(defaultProps.bookingData.checkOut);
  });

  test('should call setConfirmedBookings on submit', () => {
    const wrapper = render(setup());

    const checkOutDate = moment(defaultProps.bookingData.checkOut).add(1, 'days').format('YYYY-MM-DD')

    const inputCheckOut = wrapper.getByTestId('input-check-out') as HTMLInputElement;
    fireEvent.change(inputCheckOut, { target: { value: checkOutDate } })

    const changeDatesForm = wrapper.getByTestId("change-dates-form");

    fireEvent.submit(changeDatesForm)

    expect(defaultConfirmedBookingsContextValue.setConfirmedBookings).toHaveBeenCalled()
  });

  test('should render an error message whenever there are dates overlapping on submit', () => {
    const wrapper = render(setup({ ...defaultConfirmedBookingsContextValue, confirmedBookings: [{ ...testBookingData, id: "2" }] }));

    const changeDatesForm = wrapper.getByTestId("change-dates-form");

    fireEvent.submit(changeDatesForm)

    const errorMsg = wrapper.getByTestId("error-msg")

    expect(errorMsg).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test('should render an error message whenever the check-in and check-out are not changed on submit', () => {
    const wrapper = render(setup({ ...defaultConfirmedBookingsContextValue, confirmedBookings: [testBookingData] }));

    const changeDatesForm = wrapper.getByTestId("change-dates-form");

    fireEvent.submit(changeDatesForm)

    const errorMsg = wrapper.getByTestId("error-msg")

    expect(errorMsg).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
