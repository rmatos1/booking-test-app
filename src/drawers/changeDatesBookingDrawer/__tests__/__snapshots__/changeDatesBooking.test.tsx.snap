// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangeDatesBookingDrawer /> should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body
    style="overflow: hidden;"
  >
    <div>
      <div
        class="sc-idOjMB kWZquY"
      />
      <div
        class="sc-cxFKTC hsKcGb"
        data-testid="base-drawer"
      >
        <button
          class="sc-dmlpXa gYemGx"
          data-testid="close-button"
        >
          X
        </button>
        <form
          class="sc-crrtmM XATYL"
        >
          <h3
            class="sc-dQpIV fPWcU"
          >
            Choose the new dates
          </h3>
          <div
            class="sc-jSguLX klDGSv"
          >
            <label
              class="sc-gKsecS dRZBKP"
              data-testid="label"
            >
              Check in
            </label>
            <input
              class="sc-iBPTVF YEbKX"
              data-testid="input-check-in"
              max="2023-12-25"
              min="2023-12-29"
              name="checkIn"
              type="date"
              value="2023-12-19"
            />
          </div>
          <div
            class="sc-jSguLX klDGSv"
          >
            <label
              class="sc-gKsecS dRZBKP"
              data-testid="label"
            >
              Check out
            </label>
            <input
              class="sc-iBPTVF YEbKX"
              data-testid="input-check-out"
              min="2023-12-20"
              name="checkOut"
              type="date"
              value="2023-12-26"
            />
          </div>
          <button
            class="sc-jrAGKZ kaeceu"
            data-testid="update-booking"
          >
            Update Booking
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-idOjMB kWZquY"
    />
    <div
      class="sc-cxFKTC hsKcGb"
      data-testid="base-drawer"
    >
      <button
        class="sc-dmlpXa gYemGx"
        data-testid="close-button"
      >
        X
      </button>
      <form
        class="sc-crrtmM XATYL"
      >
        <h3
          class="sc-dQpIV fPWcU"
        >
          Choose the new dates
        </h3>
        <div
          class="sc-jSguLX klDGSv"
        >
          <label
            class="sc-gKsecS dRZBKP"
            data-testid="label"
          >
            Check in
          </label>
          <input
            class="sc-iBPTVF YEbKX"
            data-testid="input-check-in"
            max="2023-12-25"
            min="2023-12-29"
            name="checkIn"
            type="date"
            value="2023-12-19"
          />
        </div>
        <div
          class="sc-jSguLX klDGSv"
        >
          <label
            class="sc-gKsecS dRZBKP"
            data-testid="label"
          >
            Check out
          </label>
          <input
            class="sc-iBPTVF YEbKX"
            data-testid="input-check-out"
            min="2023-12-20"
            name="checkOut"
            type="date"
            value="2023-12-26"
          />
        </div>
        <button
          class="sc-jrAGKZ kaeceu"
          data-testid="update-booking"
        >
          Update Booking
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
